pkg/bpf/exporter.go

```diff
@@ -312,3 +263,72 @@ func getCPUCores() int {
 	}
 	return cores
 }
+
+type hardwarePerfEvents struct {
+	cpuCyclesPerfEvents       []int
+	cpuInstructionsPerfEvents []int
+	cacheMissPerfEvents       []int
+}
+
+func (h *hardwarePerfEvents) close() {
+	unixClosePerfEvents(h.cpuCyclesPerfEvents)
+	unixClosePerfEvents(h.cpuInstructionsPerfEvents)
+	unixClosePerfEvents(h.cacheMissPerfEvents)
+}
+
+// CreateHardwarePerfEvents creates perf events for CPU cycles, CPU instructions, and cache misses
+// and updates the corresponding eBPF maps.
+func createHardwarePerfEvents(cpuInstructionsMap, cpuCyclesMap, cacheMissMap *ebpf.Map, numCPU int) (*hardwarePerfEvents, error) {
+	var err error
+	events := &hardwarePerfEvents{
+		cpuCyclesPerfEvents:       []int{},
+		cpuInstructionsPerfEvents: []int{},
+		cacheMissPerfEvents:       []int{},
+	}
+	defer func() {
+		if err != nil {
+			unixClosePerfEvents(events.cpuCyclesPerfEvents)
+			unixClosePerfEvents(events.cpuInstructionsPerfEvents)
+			unixClosePerfEvents(events.cacheMissPerfEvents)
+		}
+	}()
+
+	// Create perf events and update each eBPF map
+	events.cpuCyclesPerfEvents, err = unixOpenPerfEvent(unix.PERF_TYPE_HARDWARE, unix.PERF_COUNT_HW_CPU_CYCLES, numCPU)
+	if err != nil {
+		klog.Warning("Failed to open perf event for CPU cycles: ", err)
+		return nil, err
+	}
+
+	events.cpuInstructionsPerfEvents, err = unixOpenPerfEvent(unix.PERF_TYPE_HARDWARE, unix.PERF_COUNT_HW_INSTRUCTIONS, numCPU)
+	if err != nil {
+		klog.Warning("Failed to open perf event for CPU instructions: ", err)
+		return nil, err
+	}
+
+	events.cacheMissPerfEvents, err = unixOpenPerfEvent(unix.PERF_TYPE_HW_CACHE, unix.PERF_COUNT_HW_CACHE_MISSES, numCPU)
+	if err != nil {
+		klog.Warning("Failed to open perf event for cache misses: ", err)
+		return nil, err
+	}
+
+	for i, fd := range events.cpuCyclesPerfEvents {
+		if err := cpuCyclesMap.Update(uint32(i), uint32(fd), ebpf.UpdateAny); err != nil {
```

Comment:
```suggestion
		if err = cpuCyclesMap.Update(uint32(i), uint32(fd), ebpf.UpdateAny); err != nil {
```
Should keep the fd open if maps fail to update?

Reply:
Nope

Reply:
Sorry .. I wasn't clear :) 
Should we keep the `fd` open if the maps fail to update?  if the answer is no, then doesn't the `if err :=` not keep it open ?

Reply:
Ahhhh yes I follow now 🤣 Fixed.